### PR TITLE
RFC Pre-create cgroup subtree of services with enabled Delegate=

### PR DESCRIPTION
**Pull request by Werkov**
This is a replacement of #22937.

The background is still the same, i.e. the conflict where we promise to give exclusive rights over the delegated subtree to the service but it can be violated when systemd needs to spawn a control command for the given service. This new approach adds a new directive `DelegateControlGroupSuffix=` where user explicitly configures subcgroup in the delegated subtree where systemd should place its processes -- that includes both the main process(es) of the service as well as control commands.

Quote from the added docs:

> It is suggested to specify `DelegateControlGroupSuffix=` with a relative path
> where systemd can put processes into the delegated subtree. The subcgroup would
> be created by systemd and service processes are moved there initially.
> The service processes are free to migrate within the hierarchy themselves.
> If you create a cgroup(s) below `DelegateControlGroupSuffix=`, be prepared that
> `ExecStartPost=`, `ExecReload=`, `ExecStop=` or `ExecStopPost=` commands may
> fail because they cannot be placed in the *inner* cgroup (i.e. this is
> discouraged).

This directive also covers possible threaded subtrees, when the user can just specify that the suffix is the top cgroup.

The use is demonstrated in `user@.service`.

Reasons for RFCness still:
* `ExecStartPre=` is handled specially (it is typically running first so there's no cgroup structure, OTOH, I don't see reasons why it should be excluded.
* I removed any tinkering with oomd attributes in the subtree, not sure if it's needed.
* I kept the original behavior with implicit `.control` cgroup for backwards compatibility. (It causes awkward config when `.` is different from `` (empty string) though.)
* The directive is only supported in the unified mode.
* It's based on failing #24958 (i.e. I plan to rebase with fixes anyway).

(Code freeze notice acknowledged, this needn't go into the release.)

### CODE REVIEW

**Comment by Werkov** [comment]
E: Unable to locate package clang-15
	E: Unable to locate package lldb-15
	E: Unable to locate package lld-15
	E: Unable to locate package clangd-15

	ERROR: test_resolved_domain_restricted_dns (__main__.DnsmasqClientTest)

Test failures seem unrelated.

File: src/core/load-fragment-gperf.gperf.in [hunk 1]
```diff
@@ -226,6 +226,7 @@
 {{type}}.TasksAccounting,                  config_parse_bool,                           0,                                  offsetof({{type}}, cgroup_context.tasks_accounting)
 {{type}}.TasksMax,                         config_parse_tasks_max,                      0,                                  offsetof({{type}}, cgroup_context.tasks_max)
 {{type}}.Delegate,                         config_parse_delegate,                       0,                                  offsetof({{type}}, cgroup_context)
+{{type}}.DelegateControlGroupSuffix,       config_parse_delegate_suffix,                0,                                  offsetof({{type}}, cgroup_context)
```
Comment (poettering) [inline src/core/load-fragment-gperf.gperf.in]: hmm, "Suffix" when it comes to file systems is usually something like ".png" or ".sh" or ".d". i think we should find a more precise name here. Maybe just "DelegateSubControlGroup"?

File: src/core/load-fragment.c [hunk 2]
```diff
@@ -3986,6 +3986,55 @@ int config_parse_delegate(
...
+        }
+
+        if (isempty(rvalue))
+                goto finish;
+
+        if (cg_all_unified() <= 0) {
+                log_syntax(unit, LOG_WARNING, filename, line, 0,
+                           "%s= setting is not supported in legacy or hybrid cgroup mode, ignoring.", lvalue);
+                return 0;
+        }
```
Comment (poettering) [inline src/core/load-fragment.c:4019]: why this restriction?

File: src/core/load-fragment.c [hunk 3]
```diff
@@ -3986,6 +3986,55 @@ int config_parse_delegate(
...
+
+        suffix = strdup(rvalue);
+
+        if (!suffix)
```
Comment (poettering) [inline src/core/load-fragment.c]: please keep function calls and their immediate error handling together, i.e. remove the empty line bove

File: src/test/test-load-fragment.c [hunk 4]
```diff
@@ -911,6 +911,7 @@ TEST(config_parse_memory_limit) {
         size_t i;
         int r;
 
+        memset(&c, 0, sizeof(c));
```
Comment (poettering) [inline src/test/test-load-fragment.c]: please initialize via `= {}` when declaring the variable.

File: src/core/load-fragment.c [hunk 5]
```diff
@@ -3986,6 +3986,55 @@ int config_parse_delegate(
...
+        assert(t != _UNIT_TYPE_INVALID);
+
+        if (!unit_vtable[t]->can_delegate ||
+            !unit_vtable[t]->exec_context_offset) {
+                log_syntax(unit, LOG_WARNING, filename, line, 0, "%s= setting not supported for this unit type, ignoring.", lvalue);
+                return 0;
+        }
+
+        if (isempty(rvalue))
+                goto finish;
```
Comment (poettering) [inline src/core/load-fragment.c]: i'd really just do 
```
if (isempty(rvalue)) {
        c->delegate_suffix = mfree(c->delegate_suffix);
        return 0;
```

freeing via free_and_replace() is just weirdn (though will work)

File: src/core/load-fragment.c [hunk 6]
```diff
@@ -3986,6 +3986,55 @@ int config_parse_delegate(
...
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+
+        _cleanup_free_ char *suffix = NULL;
+        CGroupContext *c = data;
```
Comment (poettering) [inline src/core/load-fragment.c]: `ASSERT_PTR(data)`

File: src/core/load-fragment.c [hunk 7]
```diff
@@ -3986,6 +3986,55 @@ int config_parse_delegate(
...
+                return 0;
+        }
+
+        suffix = strdup(rvalue);
+
+        if (!suffix)
+                return log_oom();
+
+        if (streq(suffix, "."))
+                suffix[0] = '\0'; /* "" */
```
Comment (poettering) [inline src/core/load-fragment.c]: hmm, why allow "." at all? and if allowing it makes sense, then why not convert it to just NULL?

**Comment by poettering** [comment]
i'd really split out the fuzz stuff into a separate PR, as it's the in the area of expertise of different people than the rest

File: src/core/execute.c [hunk 8]
```diff
@@ -5221,6 +5221,11 @@ int exec_spawn(Unit *unit,
                         if (r < 0)
                                 return log_unit_error_errno(unit, r, "Failed to create control group '%s': %m", subcgroup_path);
 
+                        r = cg_adjust_threaded(SYSTEMD_CGROUP_CONTROLLER, subcgroup_path);
+                        if (r < 0)
+                                return log_unit_error_errno(unit, r, "Failed to switch control group '%s' to threaded mode: %m",
+                                                            subcgroup_path);
+
```
Comment (poettering) [inline src/core/execute.c]: i don#t grok this. i am prety sure we should *never* create threaded cgroups. if people want those, they should do that down the tree, but not where we are affected by it.

File: docs/CGROUP_DELEGATION.md [hunk 9]
```diff
@@ -273,15 +273,6 @@ tree by the time it notifies the service manager about start-up readiness, so
...
 service manager might start `ExecStartPost=`.)
 
-(Also note, if you intend to use "threaded" cgroups — as added in Linux 4.14 —,
-then you should do that *two* levels down from the main service cgroup your
-turned delegation on for. Why that? You need one level so that systemd can
-properly create the `.control` subgroup, as described above. But that one
-cannot be threaded, since that would mean `.control` has to be threaded too —
-this is a requirement of threaded cgroups: either a cgroup and all its siblings
-are threaded or none –, but systemd expects it to be a regular cgroup. Thus you
-have to nest a second cgroup beneath it which then can be threaded.)
```
Comment (poettering) [inline docs/CGROUP_DELEGATION.md:283]: why drop this? i still think we should insist on that.

File: src/core/execute.c [hunk 10]
```diff
@@ -4014,6 +4014,7 @@ static int exec_parameters_get_cgroup_path(const ExecParameters *params, char **
          * flag, which is only passed for the former statements, not for the latter. */
 
         using_subcgroup = FLAGS_SET(params->flags, EXEC_CONTROL_CGROUP|EXEC_CGROUP_DELEGATE|EXEC_IS_CONTROL);
+        using_subcgroup &= cg_all_unified() > 0;
```
Comment (poettering) [inline src/core/execute.c]: nah, please don't. we don#t want to deviate needlessy for cgroupsv1 on this. cgroupsv1 is in maintainance mode anyway, and optimizing it is not in focus for us. Let's arrange things everywhere the same.

after all regardless if you are in cgroupsv1 or cgroupsv2 mode, it's not conceptually clear what it means to schedule a process against a cgroup, hence the "no-processes-in-inner-cgroups" rule makes sense either way.

File: src/core/service.c [hunk 11]
```diff
@@ -575,6 +575,14 @@ static int service_verify(Service *s) {
...
                                                             "Service has an empty argv in %s=. Refusing.",
                                                             service_exec_command_to_string(c));
+                        /* XXX We call cg_set_access on one level, so reject more levels that can be
+                         * configured. */
+                        if (c == SERVICE_EXEC_START &&
+                            s->cgroup_context.delegate && s->cgroup_context.delegate_suffix &&
+                            s->exec_context.user &&
+                            strchr(s->cgroup_context.delegate_suffix, '/'))
+                                return log_unit_error_errno(UNIT(s), SYNTHETIC_ERRNO(ENOEXEC),
+                                                            "DelegateControlGroupSuffix= is too deep. Refusing");
```
Comment (poettering) [inline src/core/service.c]: as mentioned earlier, I think we should not allow full path here, but single elements, already when parsing.

File: src/core/execute.h [hunk 12]
```diff
@@ -379,7 +379,7 @@ typedef enum ExecFlags {
         EXEC_NSS_DYNAMIC_BYPASS    = 1 << 5, /* Set the SYSTEMD_NSS_DYNAMIC_BYPASS environment variable, to disable nss-systemd blocking on PID 1, for use by dbus-daemon */
         EXEC_CGROUP_DELEGATE       = 1 << 6,
         EXEC_IS_CONTROL            = 1 << 7,
-        EXEC_CONTROL_CGROUP        = 1 << 8, /* Place the process not in the indicated cgroup but in a subcgroup '/.control', but only EXEC_CGROUP_DELEGATE and EXEC_IS_CONTROL is set, too */
+        EXEC_CONTROL_CGROUP        = 1 << 8, /* Place the process not in the indicated cgroup but in a subcgroup, but only when EXEC_CGROUP_DELEGATE is set, too */
```
Comment (poettering) [inline src/core/execute.h]: it's a bit of a misnomer when you apply this to the "main" process too. maybe add a separate flag for that, i.e. EXEC_MAIN_CGROUP or so?

File: src/shared/cgroup-setup.c [hunk 13]
```diff
@@ -505,6 +506,67 @@ int cg_set_access(
...
+                /* XXX paths below `top` should have access on all attributes */
+                r = cg_set_access_one(controller, path, uid, gid);
+                if (r < 0)
+                        return r;
+
+                s[n] = *p == '\0' ? '\0' : '/';
+        }
+
+        assert_not_reached();
+}
```
Comment (poettering) [inline src/shared/cgroup-setup.c:558]: i really don't like this. What do we gain by allowing full paths?

why isn't permitting a single subdir with no further depths enough?

File: src/shared/cgroup-setup.c [hunk 14]
```diff
@@ -424,26 +424,41 @@ static int cg_set_access_one(
...
+                { "cgroup.freeze",          false, false},
+                { "cgroup.kill",            false, false},
+                { "cgroup.max.depth",       false, false},
+                { "cgroup.max.descendants", false, false},
+                { "cgroup.stat",            false, false},
+                { "cgroup.type",            false, false},
+                { "cpu.pressure",           false, false},
+                { "cpu.stat",               false, false},
+                { "io.pressure",            false, false},
+                { "memory.pressure",        false, false},
```
Comment (poettering) [inline src/shared/cgroup-setup.c:435]: hmm, shouldn't we just chown everything in the subcgroup? i.e. just use recurse_dir() and fchwon() everything recursively?

File: units/user@.service.in [hunk 15]
```diff
@@ -19,6 +19,7 @@ User=%i
 PAMName=systemd-user
 Type=notify
 ExecStart={{ROOTLIBEXECDIR}}/systemd --user
+ExecReload=systemctl --user daemon-reload
```
Comment (poettering) [inline units/user@.service.in]: please don't, in light of #25916

File: units/user@.service.in [hunk 16]
```diff
@@ -23,6 +23,7 @@ ExecReload=systemctl --user daemon-reload
 Slice=user-%i.slice
 KillMode=mixed
 Delegate=pids memory cpu
+DelegateControlGroupSuffix=init.scope
```
Comment (poettering) [inline units/user@.service.in]: (btw, we should do something similar in systemd-nspawn@.service, but it's a bit more involved there, since we first need to teach nspawn to recognize if it is invoked like that. maybe leave for another PR)

**Review by poettering — CHANGES_REQUESTED** [review]
sorry for taking so long to review.

I don't like the "threaded" cgroup stuff. And I think we should create a single subdir, not a whole path of subdirs.

File: src/core/load-fragment-gperf.gperf.in (hunk 1 above)
Comment (Werkov) [reply src/core/load-fragment-gperf.gperf.in]: Name changed (also in `unit_attach_pid_to_cgroup_via_bus`).

File: src/core/load-fragment.c (hunk 7 above)
Comment (Werkov) [reply src/core/load-fragment.c]: There is difference between `DelegateSubControlCgroup=` and `DelegateSubControlCgroup=.`, former is "compat", latter no subcgroup. I'll get back to that in another comment.
 (details explained in the manpage).

File: src/core/execute.c (hunk 10 above)
Comment (Werkov) [reply src/core/execute.c]: I changed my mind and it's perhaps simpler to not special-case v1 instead of explicitly disallowing it for the subcgroup.
(Although the scheduling argument is weakened by touching only controller-less hierarchies on v1.)

File: src/core/execute.h (hunk 12 above)
Comment (Werkov) [reply src/core/execute.h]: Good catch. I've flipped the flag for easier use.

File: src/shared/cgroup-setup.c (hunk 14 above)
Comment (Werkov) [reply src/shared/cgroup-setup.c:435]: I'm not proud of this part but complete recursion won't make it because there are files that are "owned" by the delegator but are placed in the delegatee's subtree. E.g. delegatee shouldn't tune it's `memory.max` or `cgroup.max.depth` (the former example is illustrative only, subcgroup creation happens before controller enablement).
However, it's perfectly fine to modify these _any_ of the knobs deeper in the hierarchy by the delegatee.
To make the cause not so simple, _some_ 1st level knobs must actually be permitted to the delegatee (`cgroup.procs` and `cgroup.subtree_control`).
That's why this function ended up with listing all possible cgroup (core) attributes :-/

File: units/user@.service.in (hunk 15 above)
Comment (Werkov) [reply units/user@.service.in]: I see. Dropped.

File: units/user@.service.in (hunk 16 above)
Comment (Werkov) [reply units/user@.service.in]: Note taken. (Not added to current patchset.)

**Comment by Werkov** [comment]
Thanks for the review.
I've force pushed suggested little code changes and some generic suggestions.

The standing topic are:

### threaded mode
  - my goal is to allow following topologies:
```
foo.service            cgroup.type=domain threaded
  `custom-child-1  cgroup.type=threaded
  `...
  `custom-child-n  cgroup.type=threaded
```
   - in order to accommodate the compat variant with implicit `.control` wrapper, the `invalid domain`->`threaded` adjustments is needed as it's sibling of children
   - this can be avoided if compat mode was removed but that'd regress #10482
   - users should be required to provide explicit `DelegateSubControlGroup=` in such cases
   - the current form tries to cater for all (even threaded users)
   - also, I don't want to force threaded groups users to push their subtree a level deeper when it's not technically needed -- the top cgroup (`domain threaded`) is not much different from `domain` in systemd's PoV
 
### arbitrary nesting
  - my main argument is that it's more generic and nicer
    - would `systemd-nspanw@` utilize the 2nd level nesting? (`payload/init.scope`)
  - I agree that 1 level only would simplify the code (especially the chmods)

File: src/core/load-fragment.c (hunk 2 above)
Comment (Werkov) [reply src/core/load-fragment.c:4019]: I had removed the restriction earlier. (Obsolete comment now.)

**Comment by Werkov** [comment]
Rebasing force-push. (Triggered test failures[^1] unrelated to changes.)

[^1]: `E: Failed to fetch https://packages.microsoft.com/ubuntu/22.04/prod/dists/jammy/main/binary-arm64/Packages.gz` and alike... Assuming transient.

**Comment by Werkov** [comment]
I've removed the backward compat `.control` and thus the provision for threaded mode too. The arbitrary nesting remains.

File: src/core/execute.c (hunk 8 above)
Comment (poettering) [reply src/core/execute.c]: i am still not getting this. why do we bother with threaded mode? people can do this down the tree, but not up to where it affects us.

File: src/shared/cgroup-setup.c [hunk 17]
```diff
@@ -333,6 +333,32 @@ int cg_create_and_attach(const char *controller, const char *path, pid_t pid) {
...
+                return r;
+
+        if (!streq_skip_trailing_chars(contents, "domain invalid", NEWLINE))
+                return 0;
+
+        r = write_string_file(fs, "threaded", WRITE_STRING_FILE_DISABLE_BUFFER);
+        if (r < 0)
+                return r;
+        return 0;
+}
```
Comment (poettering) [inline src/shared/cgroup-setup.c]: i don't like this at all. threaded mode is something people should use down the tree, and not something we should deal with.

File: src/test/test-load-fragment.c (hunk 4 above)
Comment (poettering) [reply src/test/test-load-fragment.c]: not addressed

File: src/core/load-fragment.c [hunk 18]
```diff
@@ -3988,6 +3988,49 @@ int config_parse_delegate(
...
+        }
+
+        subcgroup = strdup(rvalue);
+        if (!subcgroup)
+                return log_oom();
+
+        if (streq(subcgroup, "."))
+                subcgroup[0] = '\0'; /* "" */
+        else if (path_simplify_and_warn(subcgroup, PATH_CHECK_RELATIVE | PATH_CHECK_FATAL, unit, filename, line, lvalue) < 0)
+                return 0;
```
Comment (poettering) [inline src/core/load-fragment.c:4026]: we should probably warn (or maybe even refuse?) about all paths that would require escaping because they can collide with names of cgroup attributes. i.e. all paths where `streq(p, cg_escape(p))` doesn't hold...

And quite frankly, I'd be conservative here, and refuse paths, and only allow filenames, i.e. single elements.

I don't see the benefit of allowing full paths here. Makes chown() logic so much harder for little gain. At least none of our own services would ever use anything more than a single level.

File: src/core/execute.c [hunk 19]
```diff
@@ -111,6 +111,10 @@
 
 #define SNDBUF_SIZE (8*1024*1024)
 
+/* Default delegated cgroup suffices */
+#define DELEGATE_CGROUP_CONTROL ".control"
+#define DELEGATE_CGROUP_PAYLOAD ""
```
Comment (poettering) [inline src/core/execute.c]: i think this doesn't make the code either to grok really. in particular the fact that DELEGATE_CGROUP_PAYLOAD is in fact *not* a valid cgroup name is utterly confusing.

macros are great for making stuff configurable at build time, or to simply complex expressions, or to make things compiler errors that are otherwise easily mistyped and never seen.

But I am not convinced either of this is the case here, and it would be a lot simpler to just spell out the group names literally, as before.

File: src/core/execute.c [hunk 20]
```diff
@@ -4022,35 +4026,40 @@ static int compile_suggested_paths(const ExecContext *c, const ExecParameters *p
...
+         * commands where the main process is already forked. For ExecStartPre= this is not necessary, the
+         * cgroup is still empty. We distinguish these cases with the EXEC_CONTROL_CGROUP flag, which is only
+         * passed for the former statements, not for the latter. */
+
+        if (FLAGS_SET(params->flags, EXEC_CONTROL_CGROUP|EXEC_CGROUP_DELEGATE|EXEC_IS_CONTROL))
+                subcgroup = c->delegate_subcgroup ?: DELEGATE_CGROUP_CONTROL;
+        else if (FLAGS_SET(params->flags, EXEC_CONTROL_CGROUP|EXEC_CGROUP_DELEGATE))
+                subcgroup = c->delegate_subcgroup ?: DELEGATE_CGROUP_PAYLOAD;
+
+        p = path_join(params->cgroup_path, subcgroup);
```
Comment (poettering) [inline src/core/execute.c:4042]: this is really hard to follow, because subcgroup can now be valid (which kinda is what most people would assume), or is NULL or is "". 

We strive to write readable code, obvious code, not super optimized code... this is not performance sensitive after all.

File: src/core/execute.c [hunk 21]
```diff
@@ -4246,6 +4255,7 @@ static int exec_child(
         int r, ngids = 0, exec_fd;
         _cleanup_free_ gid_t *supplementary_gids = NULL;
         const char *username = NULL, *groupname = NULL;
+        _cleanup_free_ char *subcgroup_path = NULL;
         _cleanup_free_ char *home_buffer = NULL;
```
Comment (poettering) [inline src/core/execute.c:4245]: merge these two lines

File: src/core/service.c [hunk 22]
```diff
@@ -2267,7 +2267,7 @@ static void service_enter_start(Service *s) {
         r = service_spawn(s,
                           c,
                           timeout,
-                          EXEC_PASS_FDS|EXEC_APPLY_SANDBOXING|EXEC_APPLY_CHROOT|EXEC_APPLY_TTY_STDIN|EXEC_SET_WATCHDOG|EXEC_WRITE_CREDENTIALS|EXEC_SETENV_MONITOR_RESULT,
+                          EXEC_PASS_FDS|EXEC_APPLY_SANDBOXING|EXEC_APPLY_CHROOT|EXEC_APPLY_TTY_STDIN|EXEC_SET_WATCHDOG|EXEC_WRITE_CREDENTIALS|EXEC_SETENV_MONITOR_RESULT|EXEC_CONTROL_CGROUP,
```
Comment (poettering) [inline src/core/service.c]: hmm? why would we start a service in the control control group? this doesn't look right

File: src/core/service.c [hunk 23]
```diff
@@ -2556,7 +2556,7 @@ static void service_run_next_main(Service *s) {
         r = service_spawn(s,
                           s->main_command,
                           s->timeout_start_usec,
-                          EXEC_PASS_FDS|EXEC_APPLY_SANDBOXING|EXEC_APPLY_CHROOT|EXEC_APPLY_TTY_STDIN|EXEC_SET_WATCHDOG|EXEC_SETENV_MONITOR_RESULT,
+                          EXEC_PASS_FDS|EXEC_APPLY_SANDBOXING|EXEC_APPLY_CHROOT|EXEC_APPLY_TTY_STDIN|EXEC_SET_WATCHDOG|EXEC_SETENV_MONITOR_RESULT|EXEC_CONTROL_CGROUP,
```
Comment (poettering) [inline src/core/service.c]: ditto

File: src/core/service.c [hunk 24]
```diff
@@ -2520,7 +2520,7 @@ static void service_run_next_control(Service *s) {
                           (IN_SET(s->control_command_id, SERVICE_EXEC_CONDITION, SERVICE_EXEC_START_PRE, SERVICE_EXEC_STOP_POST) ? EXEC_APPLY_TTY_STDIN : 0)|
                           (IN_SET(s->control_command_id, SERVICE_EXEC_STOP, SERVICE_EXEC_STOP_POST) ? EXEC_SETENV_RESULT : 0)|
                           (IN_SET(s->control_command_id, SERVICE_EXEC_START_PRE, SERVICE_EXEC_START) ? EXEC_SETENV_MONITOR_RESULT : 0)|
-                          (IN_SET(s->control_command_id, SERVICE_EXEC_START_POST, SERVICE_EXEC_RELOAD, SERVICE_EXEC_STOP, SERVICE_EXEC_STOP_POST) ? EXEC_CONTROL_CGROUP : 0),
+                          (IN_SET(s->control_command_id, SERVICE_EXEC_CONDITION, SERVICE_EXEC_START_PRE) ? EXEC_NO_SUBCGROUP : 0),
```
Comment (poettering) [inline src/core/service.c:2539]: if it's just these two types of exec lines this sounds like something we sould stop to special case.

but frankly i don't understand why it would be just these two types.

if we invoke ExecStart= with delegation on and no subcgroup set, then we should still invoke the binary in the top-level cgroup. hence ExecStart should set EXEC_NO_SUBCRGOUP too

File: units/user@.service.in [hunk 25]
```diff
@@ -22,6 +22,7 @@ ExecStart={{ROOTLIBEXECDIR}}/systemd --user
 Slice=user-%i.slice
 KillMode=mixed
 Delegate=pids memory cpu
+DelegateSubControlGroup=init.scope
```
Comment (poettering) [inline units/user@.service.in:25]: udev's create_subcgroup() logic should be reworked to make use of this too, btw.

**Comment by poettering** [comment]
> would systemd-nspanw@ utilize the 2nd level nesting? (payload/init.scope)

No. nspawn doesn't really knw what the payload will do. it's up to the payload to figure that out. only if you specify "-b" (or something equivalent), we'll even try to invoke an init system inside the container, hence quite likely we never have any further nesting.

also nspawn puts itself into "supervisor", not "payload".

**Comment by poettering** [comment]
So, I am strongly of the opinion we shouldn't bother with threaded mode. And even if I ever change my mind on that, it should not be done in the same PR as the subcgroup stuff.

Let's please dumb this down: add the delegation subdir stuff, but only that. port all three users over, so we know it works (i.e. nspawn, systemd --user, udev). Do not do deeply nested subdirs, do not do any threaded mode stuff.

File: src/shared/cgroup-setup.c (hunk 14 above)
Comment (poettering) [reply src/shared/cgroup-setup.c:435]: i don't follow?

if we define a subcgroup "foo" then everything inside of "foo" should just be chowned to user of service. Why wouldn't we?

**Comment by Werkov** [comment]
Erm, I'm not sure whether you saw the state with last two commits applied:
- service: Remove control commands' helper .control cgroup
- Revert "udev: run the main process, workers, and spawned commands in /udev subcgroup

because you seem to respond to the anterior behavior.
After these two patches:
- there is no special casing of threaded mode,
- udevd doesn't need subtree hiding.

I'll look at `systemd-nspawn@.service` in the next iteration.

**Comment by poettering** [comment]
I review PRs starting from the first patch to the last. Are you saying your first patches are not to be merged? or are undone by later patches? Then ffs, squash them, this otherwise just becomes a waste of time. Don't create humungous 22 patch PRs if major chunks of it are just stuff you immediately undo again.

We want individual commits in PRs to reflect logical steps, not your personal development history. Our tree should remain bisectable, and thus we dn't like merging PRs that contain "rubbish" commits that are backed out by later commits of the same PR.

**Comment by poettering** [comment]
Let's close this here now that #27376 has been merged. This PR covered more than the merged work, hence if there's something you still would like to see land please rebase, and file as new PRs.

That said, I am strictly against trying to support threaded cgroups managed by systemd. That's what people should do below their delegated subtrees, but they should not involve systemd in that at all.

Also, please submit them "clean", i.e. each commit should make one logical step, but not necessarily a historical one. If later commits revert work of an earlier commit that's a total no-go. Otherwise this is not reviewable.

**Comment by Werkov** [comment]
Ack. Will review it at some later time. Thanks.
